### PR TITLE
rand/randfile: use basic open() & read() in RAND_load_file()

### PR DESCRIPTION
**Pull request by sebastianas**
The recent rework in commit 9ee344f ("Cleanup
RAND_load_file,RAND_write_file") led to not accepting device nodes. It
has been said that non-regular files were never supported but a lot
projects were seeding "/dev/urandom" which got broken now.

This patch tries to fix this by using plain open() and read() and
handling the EINTR error in case a singal is sent while the reader is
busy in the syscall and the syscall has to be retried.

My VMS knowledge is non-existing and my understanding is that it is
non-UNIX standard in regard to fopen(). I've found some HP OpenVMS
documentation and based on this, open(), read() and errno (incl. EINTR)
should work as expected.

GH: #5511
Signed-off-by: Sebastian Andrzej Siewior <sebastian@breakpoint.cc>

### CODE REVIEW

**Comment by richsalz** [comment]
Please take a look at #5728 for an alternate (that I prefer, at least :)

File: crypto/rand/randfile.c [hunk 1]
```diff
@@ -76,43 +77,52 @@ int RAND_load_file(const char *file, long bytes)
...
+    ssize_t got;
+    size_t need;
 
     if (bytes == 0)
         return 0;
 
-#ifndef OPENSSL_NO_POSIX_IO
-    if (stat(file, &sb) < 0 || !S_ISREG(sb.st_mode)) {
-        RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_NOT_A_REGULAR_FILE);
+    in = open(file, O_RDONLY);
```
Comment (levitte) [inline crypto/rand/randfile.c]: On VMS, you still need the slew of little strings, something that would be easy to encode like this:

```
#define open_extra_params
#ifdef OPENSSL_SYS_VMS
# undef open_extra_params
# define open_extra_params ,"shr=get,put,upd,del","ctx=bin,stm","rfm=stm","rat=none","mrs=0"
#endif
    in = open(file, O_RDONLY, 0 open_extra_params);
```

File: crypto/rand/randfile.c [hunk 2]
```diff
@@ -76,43 +77,52 @@ int RAND_load_file(const char *file, long bytes)
...
+            need = (bytes < RAND_FILE_SIZE) ? bytes : RAND_FILE_SIZE;
         else
-            n = RAND_FILE_SIZE;
-        i = fread(buf, 1, n, in);
-        if (i <= 0)
+            need = RAND_FILE_SIZE;
+        got = read(in, buf, need);
+        if (got < 0) {
+            if (errno == EINTR)
+                continue;
```
Comment (levitte) [inline crypto/rand/randfile.c]: I'm not sure this is the right thing to do...  If it was interrupted, surely the application wants to know and decide what to do with it?
(although, looking at `rand_egd.c`, for example, this is what we usually do...  I'm undecided if that's a good thing...)

**Comment by dot-asm** [comment]
One has to keep in mind that OpenSSL doesn't make an assumption that POSIX interface is always an option. Even though it might feel appealing to use open one can't replace stdio. One can arrange it as alternative but not replace. And if you can't replace, then you either have to maintain to code paths or have combination of stdio and POSIX-y "backdoor" like fstat(fileno(fp) when available. This is the reason for why it looks the way it is and why #5737 is suggested. And naturally why modification suggested in this request is not viewed as preferred.

**Comment by sebastianas** [comment]
On 2018-03-22 23:07:39 [+0000], Richard Levitte wrote:
> levitte requested changes on this pull request.
> >  
>      if (bytes == 0)
>          return 0;
>  
> -#ifndef OPENSSL_NO_POSIX_IO
> -    if (stat(file, &sb) < 0 || !S_ISREG(sb.st_mode)) {
> -        RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_NOT_A_REGULAR_FILE);
> +    in = open(file, O_RDONLY);
> 
> On VMS, you still need the slew of little strings, something that would be easy to encode like this:
> 
> ```
> #define open_extra_params
> #ifdef OPENSSL_SYS_VMS
> # undef open_extra_params
> # define open_extra_params ,"shr=get,put,upd,del","ctx=bin,stm","rfm=stm","rat=none","mrs=0"
> #endif
>     in = open(file, O_RDONLY, 0 open_extra_params);

okay, I see. Where is this comming from? Atleast from the OpenVMS google
gave me it looks like "normal" POSIX interface. Or is this OpenVMS I
found and VMS you provide support here something different?

> 
> >          else
> -            n = RAND_FILE_SIZE;
> -        i = fread(buf, 1, n, in);
> -        if (i <= 0)
> +            need = RAND_FILE_SIZE;
> +        got = read(in, buf, need);
> +        if (got < 0) {
> +            if (errno == EINTR)
> +                continue;
> 
> I'm not sure this is the right thing to do...  If it was interrupted, surely the application wants to know and decide what to do with it?
> (although, looking at `rand_egd.c`, for example, this is what we usually do...  I'm undecided if that's a good thing...)

That is correct. EINTR means you received a signal (and handled it) and
the syscall has to be retried again. So you can *either* receive less
data because of the signal (and errno remains unchanged) *or* you
receive the signal before the syscall started doing anything so in that
case you get the error condition and errno set.

Sebastian

**Comment by levitte** [comment]
@sebastianas said thusly:
> On 2018-03-22 23:07:39 [+0000], Richard Levitte wrote:
> > levitte requested changes on this pull request.
> > 
> > On VMS, you still need the slew of little strings, something that would be easy to encode like this:
> > 
> > ```
> > #define open_extra_params
> > #ifdef OPENSSL_SYS_VMS
> > # undef open_extra_params
> > # define open_extra_params ,"shr=get,put,upd,del","ctx=bin,stm","rfm=stm","rat=none","mrs=0"
> > #endif
> >     in = open(file, O_RDONLY, 0 open_extra_params);
> > ```
> 
> okay, I see. Where is this comming from? Atleast from the OpenVMS google
> gave me it looks like "normal" POSIX interface. Or is this OpenVMS I
> found and VMS you provide support here something different?

We're talking OpenVMS alright (VMS is the same thing but got "rebranded" when DEC made the C library POSIX compliant and even created a posix shell).  And well, there's the posix compliant version of both both `open` and `fopen`, as seen [here](http://h41379.www4.hpe.com/doc/83final/5763/5763pro_050.html#index_x_1396)...  all those little strings [are explained in the `creat` section, though](http://h41379.www4.hpe.com/doc/83final/5763/5763pro_027.html#index_x_722) (and incidently, contains an example on using `open`).

The trick for googling this is "openvms c runtime library", that'll lead you to the proper (although ancient in style) manual.
> > ``` diff
> >          else
> > -            n = RAND_FILE_SIZE;
> > -        i = fread(buf, 1, n, in);
> > -        if (i <= 0)
> > +            need = RAND_FILE_SIZE;
> > +        got = read(in, buf, need);
> > +        if (got < 0) {
> > +            if (errno == EINTR)
> > +                continue;
> > ```
> > 
> > I'm not sure this is the right thing to do...  If it was interrupted, surely the application wants to know and decide what to do with it?
> > (although, looking at `rand_egd.c`, for example, this is what we usually do...  I'm undecided if that's a good thing...)
> 
> That is correct. EINTR means you received a signal (and handled it) and
> the syscall has to be retried again. So you can *either* receive less
> data because of the signal (and errno remains unchanged) *or* you
> receive the signal before the syscall started doing anything so in that
> case you get the error condition and errno set.

Point is that these are *library* routines...  it should be up to the application to decide what they want to do if an EINTR happens...  just consider the situation where a user is desperatly holding Ctrl and C down while our code is desperatly chasing after the last few needed bits of entropy?  Nooot a pretty picture.  And sure, it's a rare situation, but still....

File: crypto/rand/randfile.c [hunk 3]
```diff
@@ -12,6 +12,7 @@
 #include <errno.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <unistd.h>
```
Comment (levitte) [inline crypto/rand/randfile.c]: Did anyone already point out that Windows doesn't have `unistd.h`?

Comment (dot-asm) [reply crypto/rand/randfile.c]: > OpenSSL doesn't make an assumption that POSIX interface is always an option.

**Comment by sebastianas** [comment]
Closing in favour of #5737

**Pull request closed without merging**